indexers/src/services/hyperbridge.service.ts

```diff
@@ -1,66 +1,179 @@
-import { HyperBridgeMetrics, Transfer } from "../types/models";
-import { HyperBridgeChainMetricsService } from "./hyperbridgeChainMetrics.service";
+import {
+  PostRequestEventLog,
+  PostResponseEventLog,
+} from "../types/abi-interfaces/EthereumHostAbi";
+import { HyperBridgeStats, Relayer, Transfer } from "../types/models";
+import { HyperBridgeChainStatsService } from "./hyperbridgeChainStats.service";
+import { RelayerService } from "./relayer.service";
+import assert from "assert";
+import { isHexString } from "ethers/lib/utils";
+import { EthereumHostAbi__factory } from "../types/contracts";
 
 export class HyperBridgeService {
   /**
-   * Get the HyperBridgeMetrics entity
+   * Get the HyperBridgeStats entity
    */
-  static async getMetrics(): Promise<HyperBridgeMetrics> {
-    let metrics = await HyperBridgeMetrics.get(HYPERBRIDGE_METRICS_ENTITY_ID);
-
-    if (!metrics) {
-      metrics = HyperBridgeMetrics.create({
-        id: HYPERBRIDGE_METRICS_ENTITY_ID,
-        postRequestsHandled: BigInt(0),
-        feesPayedOut: BigInt(0),
-        feesEarned: BigInt(0),
+  static async getStats(): Promise<HyperBridgeStats> {
+    let stats = await HyperBridgeStats.get(HYPERBRIDGE_STATS_ENTITY_ID);
+
+    if (!stats) {
+      stats = HyperBridgeStats.create({
+        id: HYPERBRIDGE_STATS_ENTITY_ID,
+        numberOfMessagesSent: BigInt(0),
+        numberOfSuccessfulMessagesSent: BigInt(0),
+        numberOfTimedOutMessages: BigInt(0),
+        numberOfUniqueRelayers: BigInt(0),
+        feesPayedOutToRelayers: BigInt(0),
+        protocolFeesEarned: BigInt(0),
         totalTransfersIn: BigInt(0),
       });
 
-      await metrics.save();
+      await stats.save();
     }
 
-    return metrics;
+    return stats;
   }
 
   /**
-   * Increment the number of post requests handled by hyperbridge
+   * Perform the necessary actions related to Hyperbridge stats when a PostRequest/PostResponse event is indexed
    */
-  static async incrementNumberOfPostRequestsHandled(
+  static async handlePostRequestOrResponseEvent(
     chain: SupportedChain,
+    event: PostRequestEventLog | PostResponseEventLog,
   ): Promise<void> {
-    let metrics = await this.getMetrics();
-    metrics.postRequestsHandled += BigInt(1);
+    assert(
+      event.args,
+      "No handlePostRequestEvent/handlePostResponseEvent args",
+    );
 
-    // Update the specific chain metrics
-    let chainMetrics =
-      await HyperBridgeChainMetricsService.findOrCreateChainMetrics(chain);
-    chainMetrics.postRequestsHandled += BigInt(1);
+    const { args, address } = event;
+    let { data } = args;
+
+    const protocolFee = await this.computeProtocolFeeFromHexData(address, data);
+
+    await this.incrementProtocolFeesEarned(protocolFee, chain);
+  }
+
+  /**
+   * Perform the necessary actions related to Hyperbridge stats when a PostRequestHandled/PostResponseHandled event is indexed
+   */
+  static async handlePostRequestOrResponseHandledEvent(
```

Comment:
There will be no PostRequestHandled or PostResponseHandled events for failed transactions, that's why we need to index `handlePostRequest` and `handlePostResponse` transactions.

Reply:
Done, the metric calculation for failed messages is now done on the `handlePostRequest` and `handlePostResponse` transactions.